lib/transport/transport.go

```diff
@@ -0,0 +1,205 @@
+	cancel  context.CancelFunc
+	err     error
+	errLock sync.Mutex
+}
+
+func (t *subscription) Events() EventChan {
+	return EventChan(t.events)
+}
+
+func (t *subscription) Close() {
+	t.cancel()
+}
+
+func (t *subscription) Err() error {
+	t.errLock.Lock()
+	defer t.errLock.Unlock()
+	return t.err
+}
+
+func Subscribe(masterURI string, fi *mesosproto.FrameworkInfo, force bool) (Subscription, error) {
+	call := mesosproto.Call{
+		Type: mesosproto.Call_SUBSCRIBE.Enum(),
+		Subscribe: &mesosproto.Call_Subscribe{
+			FrameworkInfo: fi,
+			Force: &force,
+		},
+	}
+
+	body, err := proto.Marshal(&call)
+	if err != nil {
+		return nil, err
+	}
+
+	const EP_SCHEDULER = "/master/api/v1/scheduler"
+	req, err := http.NewRequest("POST", "http://"+masterURI+EP_SCHEDULER, bytes.NewReader(body))
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Set("Content-Type", "application/x-protobuf")
+	req.Header.Set("Accept", "application/json")
+
+	var t *subscription
+	bg := context.Background()
+	ctx, _ := context.WithTimeout(bg, 75*time.Second)
+	err = httpDo(ctx, req, func(resp *http.Response, err error) error {
+		if err != nil {
+			return err
+		}
+		if resp.StatusCode != 200 {
+			defer resp.Body.Close()
+			body, err := ioutil.ReadAll(resp.Body)
+			if err != nil {
+				return err
+			}
+			return fmt.Errorf("%s", body)
+		}
+		c2, cancel := context.WithCancel(bg)
+		t = &subscription{
+			body:   resp.Body,
+			events: make(chan *mesosproto.Event, 100),
+			cancel: cancel,
+		}
+		go t.handleEvents(c2)
+		return nil
+	})
+	return t, err
+}
+
+func httpDo(ctx context.Context, req *http.Request, f func(*http.Response, error) error) error {
+	// Run the HTTP request in a goroutine and pass the response to f.
+	tr := &http.Transport{}
+	client := &http.Client{Transport: tr}
+	c := make(chan error, 1)
+	go func() { c <- f(client.Do(req)) }()
+	select {
+	case <-ctx.Done():
+		tr.CancelRequest(req)
+		<-c // Wait for f to return.
+		return ctx.Err()
+	case err := <-c:
+		return err
+	}
+}
+
+func (t *subscription) setError(err error) {
+	t.errLock.Lock()
+	defer t.errLock.Unlock()
+	t.err = err
+}
+
+func (t *subscription) handleEvents(ctx context.Context) {
+	defer t.cancel()
+	defer t.body.Close()
+	defer close(t.events)
+
+	var empty bool
+	un := records.Unmarshaler(func(b []byte, v interface{}) error {
+		if b == nil || len(b) == 0 {
+			empty = true
+			return nil
+		}
+		return json.Unmarshal(b, v)
+	})
+	errNoPulse := errors.New("failed to receive a pulse")
+	dec := records.NewDecoder(t.body, un)
+	timed := timedDecoder(dec, 15*time.Second, 5, errNoPulse)
+	for {
+		empty = false
+		event := &mesosproto.Event{}
+		if err := timed.Decode(event); err != nil {
+			if err == io.EOF || err == errNoPulse {
+				t.setError(err)
+				return
+			}
+			// protocol error? log and attempt to continue
+			// TODO(jdef) not all protocol errors are recoverable
+			// TODO(jdef) log all of these
+			log.Println("ERROR:", err)
+			continue
+		}
+		if empty {
```

Comment:
vinod has updated the spec: this should never happen. should log a warning and continue